### PR TITLE
feat(spanner): CMEK backup and restore

### PR DESCRIPTION
**Pull request by mr-salty**
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5761)

### CODE REVIEW

**Comment by mr-salty** [comment]
FWIW, I haven't actually run the integration test yet, and I'm not sure if it will pass in CI or not. Let's sync up about that tomorrow.

File: google/cloud/spanner/integration_tests/backup_integration_test.cc [hunk 1]
```diff
@@ -221,6 +228,83 @@ TEST_F(BackupTestWithCleanup, BackupTestSuite) {
...
+  auto instance_id =
+      spanner_testing::PickRandomInstance(generator, project_id_);
+  ASSERT_STATUS_OK(instance_id);
+  std::string database_id =
+      google::cloud::spanner_testing::RandomDatabaseName(generator);
+  std::string restore_database_id =
+      google::cloud::spanner_testing::RandomDatabaseName(generator);
+  ASSERT_FALSE(project_id_.empty());
+  Instance in(project_id_, *instance_id);
+  Database db(project_id_, *instance_id, database_id);
```
Comment (devbww) [inline google/cloud/spanner/integration_tests/backup_integration_test.cc:246]: `Database db(in, database_id)` perhaps?

File: google/cloud/spanner/database_admin_connection_test.cc [hunk 2]
```diff
@@ -476,7 +509,49 @@ TEST(DatabaseAdminClientTest, RestoreDatabaseSuccess) {
...
+        op.set_done(true);
+        gcsa::Database database;
+        database.set_name("test-db");
+        op.mutable_response()->PackFrom(database);
+        return make_status_or(op);
+      });
+
+  auto conn = CreateTestingConnection(std::move(mock));
+  Database dbase("test-project", "test-instance", "test-db");
+  Backup backup(Instance("test-project", "test-instance"), "test-backup");
```
Comment (devbww) [inline google/cloud/spanner/database_admin_connection_test.cc:551]: `Backup backup(dbase.instance(), "test-backup")` perhaps?

**Review by devbww — COMMENTED** [review]
> FWIW, I haven't actually run the integration test yet, and I'm not sure if it will pass in CI or not. Let's sync up about that tomorrow.

Ack.

**Review by mr-salty — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5761)* status: 0 of 6 files reviewed, 2 unresolved discussions (waiting on @devbww)

___
*[google/cloud/spanner/database_admin_connection_test.cc, line 551 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5761#-MS6z7Hpr7HS4gFM0rnL-r1-551:-MSBa5YHAE78Ude7519w:bo2jrji) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/1e11d531202d3f225dab083b95440f8706917ef1/google/cloud/spanner/database_admin_connection_test.cc#L551)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

`Backup backup(dbase.instance(), "test-backup")` perhaps?
</blockquote></details>

Done (note the existing `*RestoreDatabase*`tests could use the same treatment but I didn't want to do that in this CL since it's not CMEK).
___
*[google/cloud/spanner/integration_tests/backup_integration_test.cc, line 246 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5761#-MS6z7Hpr7HS4gFM0rnM-r1-246:-MSBbEQ26I1UrcyFcgA2:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/1e11d531202d3f225dab083b95440f8706917ef1/google/cloud/spanner/integration_tests/backup_integration_test.cc#L246)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

`Database db(in, database_id)` perhaps?
</blockquote></details>

Done.

**Review by devbww — COMMENTED** [review]
> Reviewed 4 of 6 files at r1, 2 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5761)* status: :shipit: complete! all files reviewed, all discussions resolved

**Review by mr-salty — COMMENTED** [review]
The integration test passes against staging spanner (it's not yet enabled in prod), you need to set:

```
export GOOGLE_CLOUD_CPP_SPANNER_SLOW_INTEGRATION_TESTS="backup"
export GOOGLE_CLOUD_CPP_SPANNER_DEFAULT_ENDPOINT="staging-wrenchworks.sandbox.googleapis.com"
```

it's quite slow against staging although bww thought it may be tolerable when enabled in prod based on his PITR testing.

```
//google/cloud/spanner/integration_tests:backup_integration_test         PASSED in 585.5s
```

> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5761#-:-MSC4_9S343SIh1TF-Vc:bjoy5y7)* status: 4 of 6 files reviewed, all discussions resolved (waiting on @devbww)

**Review by devbww — APPROVED** [review]
> Reviewed 4 of 4 files at r3.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5761)* status: :shipit: complete! all files reviewed, all discussions resolved

**Pull request merged**